File: python_solutions/1236.web-crawler.cpp
// Containers
#include <deque>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include "conf.d/say.h"

using namespace std;
int dirs[5] = {-1, 0, 1, 0, -1};

// --------------------------------------------------------
/**
 * // This is the HtmlParser's API interface.
 * // You should not implement it, or speculate about its implementation
 * class HtmlParser {
 *   public:
 *     vector<string> getUrls(string url);
 * };
 */

class Solution {
public:
  vector<string> crawl(string startUrl, HtmlParser htmlParser) {
    std::unordered_set<string> visited;
    visited.insert(startUrl);
    std::queue<string> todos;
    todos.push(startUrl);
    string host = startUrl.substr(0, startUrl.find("/", 7));

    while (!todos.empty()) {
      auto cur = todos.front();
      todos.pop();
      for (const auto &nxt : htmlParser.getUrls(cur)) {
        if (nxt.find(host) == 0 && visited.count(nxt) == 0) {
          visited.insert(nxt);
          todos.push(nxt);
        }
      }
    }

    return std::vector<string>(visited.begin(), visited.end());
  }
};

int main() { Solution s; }

File: python_solutions/1236.web-crawler.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right 
from functools import reduce, lru_cache 
from typing import List 
import itertools 
import math 
import heapq 
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right 
from functools import reduce, lru_cache 
from typing import List 
import itertools 
import math, re
import heapq 
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007



# https://leetcode.com/problems/web-crawler/description/
# Medium
# """
# This is HtmlParser's API interface.
# You should not implement it, or speculate about its implementation
# """
#class HtmlParser(object):
#    def getUrls(self, url):
#        """
#        :type url: str
#        :rtype List[str]
#        """

class Solution:
    def crawl(self, startUrl: str, htmlParser: 'HtmlParser') -> List[str]:
        visited = {}
        visited[startUrl] = True
        q = deque()
        q.append(startUrl)
        domain = re.findall(r'http://(.*?)/', startUrl + "/")[0]

        while q:
            url = q.popleft()
            for nxt in htmlParser.getUrls(url):
                if nxt.startswith(f'http://{domain}') and nxt not in visited:
                    visited[nxt] = True
                    q.append(nxt)
        return list(visited)

url = "http://news.yahoo.com/news/topics/"
domain = re.findall(r'http://(.*?)/', url)
print(domain)